### ISSUE TITLE
Subquery not recognised

### ISSUE THREAD

**Issue by ondrejmalina** [sqlfluff/sqlfluff#6176]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

If I lint/fix a file, it does not recognize a subquery.

### Expected Behaviour

SQLFluff identifies a subquery and fixes it or warns about it.

### Observed Behaviour

SQLFluff simply does not see the issue. All other issues in the code are recognised and fixed correctly. Subqueries are recognised in different, more complex files (in the same dialect).

### pyproject.toml
```toml
[tool.sqlfluff.core]
dialect = "bigquery"

[tool.sqlfluff.rules.capitalisation.keywords]
capitalisation_policy = "upper"

[tool.sqlfluff.rules.capitalisation.functions]
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.structure.subquery]
forbid_subquery_in = "both"
```

### sql_file.sql
```sql
select
    d AS days
from (
    select *
    from
        UNNEST(
            GENERATE_DATE_ARRAY(
                '2016-01-01',
                DATE_ADD(DATE_TRUNC(CURRENT_DATE(), YEAR), INTERVAL 10 YEAR),
                INTERVAL 1 DAY
            )
        ) as d
)

order by 1
```

### cli command
```bash
sqlfluff lint sql_file.sql
```

### Dialect

bigquery

### Version

Python = 3.9.2
SQLFluff = 3.1.1

### pyproject.toml
```toml
[tool.sqlfluff.core]
dialect = "bigquery"

[tool.sqlfluff.rules.capitalisation.keywords]
capitalisation_policy = "upper"

[tool.sqlfluff.rules.capitalisation.functions]
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.structure.subquery]
forbid_subquery_in = "both"
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
The above statement parser properly with the latest `sqlfluff`

**Comment by keraion** [comment]
The issue isn't the parsing, but rather ST05 doesn't trigger on the subquery.

**Comment by peterbud** [comment]
> The issue isn't the parsing, but rather ST05 doesn't trigger on the subquery.

Now I see what you mean!
